flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -741,6 +763,18 @@ private void establishJobManagerConnection(JobID jobId, JobMasterGateway jobMast
 			jobManagerTable.put(jobId, associateWithJobManager(jobMasterGateway, jobManagerLeaderId, registrationSuccess.getBlobPort()));
 		}
 
+		heartbeatManager.monitorTarget(registrationSuccess.getResourceID(), new HeartbeatTarget() {
+			@Override
+			public void sendHeartbeat(ResourceID resourceID, Object payload) {
+				jobMasterGateway.heartbeatFromTaskManager(resourceID, payload);
+			}
+
+			@Override
+			public void requestHeartbeat(ResourceID resourceID, Object payload) {
+				throw new UnsupportedOperationException("Should never call requestHeartbeat in task manager.");
```

Comment:
A heartbeat request is totally fine. This should not throw an `UnsupportedOperationException` but trigger a heartbeat being sent back to `resourceID`.